File: src/Lavoisier/Provider/Xai/Responses.hs
{-# LANGUAGE DataKinds #-}

{- | xAI's __Responses API__ (@POST \/v1\/responses@) — the Agent-Tools transport.

A third xAI transport beside "Lavoisier.Provider.Xai" (@chat\/completions@) and
"Lavoisier.Provider.Xai.Grpc", not a replacement for either. It exists because xAI's provider-run
tools live only here: the old route into them, Live Search via @search_parameters@ on
@chat\/completions@, has returned @410 Gone@ since 2026-01-12.

The request is a different shape from @chat\/completions@ — @input@ rather than @messages@,
@instructions@ rather than a system message, @max_output_tokens@ rather than @max_tokens@, and tools
declared __flat__ (@{"type":"function","name":…}@) rather than nested under @"function"@. The
response stream is decoded by "Lavoisier.Provider.Xai.ResponsesSse".
-}
module Lavoisier.Provider.Xai.Responses (
    xaiResponsesFromEnv,
    xaiResponsesProvider,
    XaiResponsesConfig (..),
    newXaiResponsesConfig,
    defaultResponsesBaseUrl,

    -- * exposed for testing
    XaiResponsesCaps,
    buildBody,
    inputItems,
    serverToolJson,
)
where

import Control.Exception (SomeException, try)
import Data.Aeson (Value (..), encode, object, toJSON, (.=))
import Data.IORef (newIORef, readIORef, writeIORef)
import Data.Maybe (mapMaybe)
import Data.Text (Text)
import Data.Text.Encoding (decodeUtf8Lenient, encodeUtf8)
import Network.HTTP.Client
import Network.HTTP.Client.TLS (tlsManagerSettings)
import Network.HTTP.Types (RequestHeaders)
import Network.HTTP.Types.Status (statusCode, statusIsSuccessful)
import System.Environment (lookupEnv)

import Data.ByteString qualified as BS
import Data.ByteString.Lazy qualified as BL
import Data.Text qualified as T

import Lavoisier.Domain (ModelId (..), renderDate)
import Lavoisier.Protocol.Message
import Lavoisier.Protocol.Provider
import Lavoisier.Protocol.Stream (Producer (..))
import Lavoisier.Provider.Xai.ResponsesSse (initResp, respEof, respPush)

-- | Connection settings. @xrBaseUrl@ has no trailing @\/responses@ — that is appended.
data XaiResponsesConfig = XaiResponsesConfig
    { xrApiKey ∷ Text
    , xrBaseUrl ∷ Text
    , xrManager ∷ Manager
    }

-- | @https:\/\/api.x.ai\/v1@.
defaultResponsesBaseUrl ∷ Text
defaultResponsesBaseUrl = "https://api.x.ai/v1"

newXaiResponsesConfig ∷ Text → Text → IO XaiResponsesConfig
newXaiResponsesConfig apiKey baseUrl = do
    mgr ← newManager tlsManagerSettings
    pure (XaiResponsesConfig apiKey baseUrl mgr)

-- | Construct from @XAI_API_KEY@ (required) and @XAI_BASE_URL@ (optional).
xaiResponsesFromEnv ∷ IO (Either ProviderError Provider)
xaiResponsesFromEnv = do
    mkey ← lookupEnv "XAI_API_KEY"
    case mkey of
        Nothing → pure (Left (PConfig "XAI_API_KEY is not set"))
        Just key → do
            base ← maybe defaultResponsesBaseUrl T.pack <$> lookupEnv "XAI_BASE_URL"
            cfg ← newXaiResponsesConfig (T.pack key) base
            pure (Right (xaiResponsesProvider cfg))

{- | Everything this transport supports, named once so 'declare' and 'negotiate' cannot drift apart.

Deliberately absent, each for a reason:

* __'PromptCaching'__ — xAI caches server-side with no request markers, and the usage payload
  carries @input_tokens_details.cached_tokens@ but no cache-/creation/ counter.
* __'Vision'__, __'StructuredOutput'__, __'StopSequences'__ — the request fields exist in the
  OpenAI-shaped API but were not exercised against the live endpoint, so declaring them would let a
  request through on an unverified encoding.
* __'TopK'__ — xAI has no @top_k@ on any transport.
* __'WebFetch'__, __'UrlContext'__, __'ClientBuiltinTools'__, __'RemoteMcp'__ — other providers\'
  tools.
-}
type XaiResponsesCaps =
    '[ 'ExtendedThinking
     , 'Sampling
     , 'ToolChoiceControl
     , 'WebSearch
     , 'XSearch
     , 'CodeExecution
     , 'CollectionsSearch
     ]

xaiResponsesProvider ∷ XaiResponsesConfig → Provider
xaiResponsesProvider cfg =
    Provider
        { providerStream = responsesStream cfg
        , providerCapabilities = declare @XaiResponsesCaps
        , providerCountTokens = \_ → pure (Right Nothing)
        }

-- --- streaming -------------------------------------------------------------------------------------

responsesStream ∷ XaiResponsesConfig → ChatRequest → IO (Either ProviderError EventStream)
responsesStream cfg chatReq = withNegotiated @XaiResponsesCaps chatReq $ \nreq → do
    let body = encode (buildBody nreq)
        url = T.unpack (T.dropWhileEnd (== '/') (xrBaseUrl cfg)) <> "/responses"
    ereq0 ← try (parseRequest url) ∷ IO (Either SomeException Request)
    case ereq0 of
        Left e → pure (Left (PConfig (T.pack (show e))))
        Right base → do
            let httpReq =
                    base
                        { method = "POST"
                        , requestHeaders = headers cfg
                        , requestBody = RequestBodyLBS body
                        , responseTimeout = responseTimeoutNone
                        }
            eresp ← try (responseOpen httpReq (xrManager cfg)) ∷ IO (Either HttpException (Response BodyReader))
            case eresp of
                Left e → pure (Left (PTransport (T.pack (show e))))
                Right resp
                    | statusIsSuccessful (responseStatus resp) → Right <$> makeStream resp
                    | otherwise → do
                        errBody ← BS.concat <$> brConsume (responseBody resp)
                        responseClose resp
                        pure (Left (PApi (statusCode (responseStatus resp)) (decodeUtf8Lenient errBody)))

headers ∷ XaiResponsesConfig → RequestHeaders
headers cfg =
    [ ("content-type", "application/json")
    , ("authorization", "Bearer " <> encodeUtf8 (xrApiKey cfg))
    ]

-- | Wrap the open response in a pull 'Producer' over the Responses SSE decoder.
makeStream ∷ Response BodyReader → IO EventStream
makeStream resp = do
    stateRef ← newIORef initResp
    pendRef ← newIORef []
    doneRef ← newIORef False
    let feed [] = pull
        feed (x : xs) = writeIORef pendRef xs >> pure (Just x)
        pull =
            readIORef pendRef >>= \case
                (x : xs) → writeIORef pendRef xs >> pure (Just x)
                [] →
                    readIORef doneRef >>= \case
                        True → pure Nothing
                        False → do
                            chunk ← brRead (responseBody resp)
                            if BS.null chunk
                                then do
                                    st ← readIORef stateRef
                                    responseClose resp
                                    writeIORef doneRef True
                                    feed (respEof st)
                                else do
                                    st ← readIORef stateRef
                                    let (st', evs) = respPush st chunk
                                    writeIORef stateRef st'
                                    feed evs
    pure (Producer pull)

-- --- the request body ------------------------------------------------------------------------------

-- | The @\/v1\/responses@ body.
buildBody ∷ Negotiated caps → Value
buildBody nreq =
    object $
        [ "model" .= unModelId (crModel req)
        , "stream" .= True
        , "input" .= inputItems (crMessages req)
        , "max_output_tokens" .= crMaxTokens req
        ]
            <> ["instructions" .= spText sp | Just sp ← [crSystem req]]
            <> ["temperature" .= t | Just t ← [crTemperature req]]
            <> ["top_p" .= p | Just p ← [crTopP req]]
            <> ["reasoning" .= object ["effort" .= effortOf l] | Just l ← [crThinking req], l /= ThinkOff]
            <> ["tool_choice" .= toolChoiceJson tc | Just tc ← [crToolChoice req]]
            <> ["tools" .= allTools | not (null allTools)]
    where
        req = negotiatedRequest nreq
        allTools = map functionToolJson (crTools req) <> mapMaybe serverToolJson (crServerTools req)

{- | @low@ or @high@ — xAI's only two reasoning efforts (medium rounds down, not up: on a
cost-weighted metric, buying more reasoning than was asked for is the worse error).
-}
effortOf ∷ ThinkingLevel → Text
effortOf = \case
    ThinkHigh → "high"
    _ → "low"

{- | A client-side tool. Responses declares these __flat__ — @{"type":"function","name":…}@ — where
@chat\/completions@ nests them under @"function"@. Verified against the live endpoint.
-}
functionToolJson ∷ ToolDef → Value
functionToolJson t =
    object
        [ "type" .= ("function" ∷ Text)
        , "name" .= tdName t
        , "description" .= tdDescription t
        , "parameters" .= tdSchema t
        ]

{- | A provider-run tool. 'Nothing' for tools this transport does not run; 'negotiate' refuses those
before they reach here, and a test pins the two lists together.
-}
serverToolJson ∷ ServerTool → Maybe Value
serverToolJson = \case
    STWebSearch _ allowed blocked →
        Just . object $
            ["type" .= ("web_search" ∷ Text)]
                -- allowed_domains and excluded_domains cannot both be set; allowed wins.
                <> [ "filters" .= object ["allowed_domains" .= allowed]
                   | not (null allowed)
                   ]
                <> [ "filters" .= object ["excluded_domains" .= blocked]
                   | null allowed && not (null blocked)
                   ]
    STXSearch allowed blocked from to →
        Just . object $
            ["type" .= ("x_search" ∷ Text)]
                <> ["allowed_x_handles" .= allowed | not (null allowed)]
                <> ["excluded_x_handles" .= blocked | null allowed, not (null blocked)]
                <> ["from_date" .= renderDate d | Just d ← [from]]
                <> ["to_date" .= renderDate d | Just d ← [to]]
    STCodeExecution → Just (object ["type" .= ("code_interpreter" ∷ Text)])
    STCollectionsSearch ids lim →
        Just . object $
            ["type" .= ("collections_search" ∷ Text), "collection_ids" .= ids]
                <> ["limit" .= n | Just n ← [lim]]
    STWebFetch {} → Nothing
    STUrlContext → Nothing

toolChoiceJson ∷ ToolChoice → Value
toolChoiceJson = \case
    ChoiceAuto → String "auto"
    ChoiceRequired → String "required"
    ChoiceNone → String "none"
    ChoiceTool n → object ["type" .= ("function" ∷ Text), "name" .= n]

{- | Messages → the @input@ array.

Assistant tool calls and their results round-trip as @function_call@ \/ @function_call_output@ items
keyed by @call_id@ — the id the stream reports as @call_id@, __not__ the @fc_…@ item id.

Verified live on 2026-08-25 by a three-turn agent run: the model called a tool, saw its result, and
answered from it, which only happens if both the @function_call@ echo and the
@function_call_output@ item are encoded as the provider expects.
-}
inputItems ∷ [Message] → Value
inputItems msgs = toJSON (concatMap one msgs)
    where
        one m = concatMap (blockItem (roleText (msgRole m))) (msgContent m)

        blockItem role = \case
            TextBlock t _ | not (T.null t) → [object ["role" .= role, "content" .= t]]
            ThinkingBlock _ → []
            ToolUseBlock callId name args →
                [ object
                    [ "type" .= ("function_call" ∷ Text)
                    , "call_id" .= callId
                    , "name" .= name
                    , "arguments" .= encodeArgs args
                    ]
                ]
            ToolResultBlock callId content _ →
                [ object
                    [ "type" .= ("function_call_output" ∷ Text)
                    , "call_id" .= callId
                    , "output" .= content
                    ]
                ]
            _ → []

        roleText ∷ Role → Text
        roleText = \case
            User → "user"
            Assistant → "assistant"

        encodeArgs ∷ Value → Text
        encodeArgs v = decodeUtf8Lenient (BL.toStrict (encode v))

File: crates/lvz-xai/src/responses.rs
//! xAI's **Responses API** (`POST /v1/responses`) — the Agent-Tools transport.
//!
//! A third xAI transport beside [`http`](crate::http) (`chat/completions`) and
//! [`grpc`](crate::grpc), not a replacement for either. It exists because xAI's provider-run tools
//! live only here: the old route into them, Live Search via `search_parameters` on
//! `chat/completions`, has returned **410 Gone since 2026-01-12**.
//!
//! The request is a different shape from `chat/completions` — `input` rather than `messages`,
//! `instructions` rather than a system message, `max_output_tokens` rather than `max_tokens`, and
//! tools declared **flat** (`{"type":"function","name":…}`) rather than nested under `"function"`.
//! The response stream is decoded by [`responses_sse`](crate::responses_sse).
//!
//! Note that `max_output_tokens` does **not** bound reasoning tokens — a 512-token ceiling has been
//! observed returning 1273 output tokens. The field is sent; the provider does not honour it as a
//! ceiling. Budget accordingly.

use std::collections::VecDeque;

use async_trait::async_trait;
use bytes::Bytes;
use futures::stream::{self, BoxStream, StreamExt};
use lvz_protocol::{
    retry_transient, with_negotiated, Capabilities, Capability, ChatRequest, ContentBlock, Event,
    Message, Negotiated, Provider, ProviderCaps, ProviderError, Role, ServerTool, ThinkingLevel,
    ToolChoice, ToolDef,
};
use serde_json::{json, Value};

use crate::responses_sse::RespDecoder;

pub(crate) const DEFAULT_BASE_URL: &str = "https://api.x.ai/v1";

/// A [`Provider`] backed by xAI's Responses API.
pub struct ResponsesTransport {
    api_key: String,
    base_url: String,
    http: reqwest::Client,
}

impl ResponsesTransport {
    /// Construct against the default base URL (`https://api.x.ai/v1`).
    pub fn new(api_key: impl Into<String>) -> Self {
        Self::with_base_url(api_key, DEFAULT_BASE_URL)
    }

    /// Construct against an explicit base URL (e.g. a proxy or a mock server in tests). The URL
    /// carries no trailing `/responses` — that is appended.
    pub fn with_base_url(api_key: impl Into<String>, base_url: impl Into<String>) -> Self {
        Self {
            api_key: api_key.into(),
            base_url: base_url.into(),
            http: reqwest::Client::new(),
        }
    }

    /// Construct from `XAI_API_KEY` (required) and `XAI_BASE_URL` (optional).
    pub fn from_env() -> Result<Self, ProviderError> {
        let api_key = std::env::var("XAI_API_KEY")
            .map_err(|_| ProviderError::Config("XAI_API_KEY is not set".into()))?;
        let base_url = std::env::var("XAI_BASE_URL").unwrap_or_else(|_| DEFAULT_BASE_URL.into());
        Ok(Self::with_base_url(api_key, base_url))
    }
}

/// Everything this transport supports, named once so the declaration and the check cannot drift.
///
/// Deliberately absent, each for a reason:
///
/// * **`PromptCaching`** — xAI caches server-side with no request markers, and the usage payload
///   carries `input_tokens_details.cached_tokens` but no cache-*creation* counter.
/// * **`Vision`**, **`StructuredOutput`**, **`StopSequences`** — the request fields exist in the
///   OpenAI-shaped API but have not been exercised against the live endpoint, so declaring them
///   would let a request through on an unverified encoding.
/// * **`TopK`** — xAI has no `top_k` on any transport.
/// * **`WebFetch`**, **`UrlContext`**, **`ClientBuiltinTools`**, **`RemoteMcp`** — other providers'
///   tools.
pub struct XaiResponsesCaps;

impl ProviderCaps for XaiResponsesCaps {
    const CAPS: &'static [Capability] = &[
        Capability::ExtendedThinking,
        Capability::Sampling,
        Capability::ToolChoiceControl,
        Capability::WebSearch,
        Capability::XSearch,
        Capability::CodeExecution,
        Capability::CollectionsSearch,
    ];
}

#[async_trait]
impl Provider for ResponsesTransport {
    /// Negotiate, then send: this method is the negotiation call and nothing else.
    async fn stream(
        &self,
        req: ChatRequest,
    ) -> Result<BoxStream<'static, Result<Event, ProviderError>>, ProviderError> {
        with_negotiated::<XaiResponsesCaps, _, _>(req, |nreq| self.send(nreq)).await
    }

    fn capabilities(&self) -> Capabilities {
        XaiResponsesCaps::declare()
    }
}

impl ResponsesTransport {
    async fn send(
        &self,
        nreq: Negotiated<XaiResponsesCaps>,
    ) -> Result<BoxStream<'static, Result<Event, ProviderError>>, ProviderError> {
        let body = build_body(&nreq);
        let url = format!("{}/responses", self.base_url.trim_end_matches('/'));

        let max_retries: u32 = std::env::var("XAI_MAX_RETRIES")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(6);
        let resp = retry_transient(max_retries, || async {
            let resp = self
                .http
                .post(&url)
                .bearer_auth(&self.api_key)
                .json(&body)
                .send()
                .await
                .map_err(|e| ProviderError::Transport(e.to_string()))?;
            let status = resp.status();
            if status.is_success() {
                Ok(resp)
            } else {
                Err(ProviderError::Api {
                    status: status.as_u16(),
                    message: resp.text().await.unwrap_or_default(),
                })
            }
        })
        .await?;

        let state = RespState {
            body: resp.bytes_stream().boxed(),
            decoder: RespDecoder::default(),
            pending: VecDeque::new(),
            drained: false,
        };

        let events = stream::unfold(state, |mut st| async move {
            loop {
                if let Some(event) = st.pending.pop_front() {
                    return Some((event, st));
                }
                if st.drained {
                    return None;
                }
                match st.body.next().await {
                    Some(Ok(chunk)) => st.decoder.push(&chunk, &mut st.pending),
                    Some(Err(e)) => {
                        st.pending
                            .push_back(Err(ProviderError::Transport(e.to_string())));
                        st.drained = true;
                    }
                    None => {
                        st.decoder.eof(&mut st.pending);
                        st.drained = true;
                    }
                }
            }
        });

        Ok(events.boxed())
    }
}

struct RespState {
    body: BoxStream<'static, reqwest::Result<Bytes>>,
    decoder: RespDecoder,
    pending: VecDeque<Result<Event, ProviderError>>,
    drained: bool,
}

// --- the request body ---

/// Build the `/v1/responses` body from a negotiated request.
pub(crate) fn build_body(nreq: &Negotiated<XaiResponsesCaps>) -> Value {
    let req = nreq.request();
    let mut body = json!({
        "model": req.model,
        "stream": true,
        "input": input_items(&req.messages),
        "max_output_tokens": req.max_tokens,
    });
    if let Some(sp) = &req.system {
        body["instructions"] = json!(sp.text);
    }
    if let Some(t) = req.temperature {
        body["temperature"] = json!(t);
    }
    if let Some(p) = req.top_p {
        body["top_p"] = json!(p);
    }
    if let Some(l) = req.thinking {
        if l != ThinkingLevel::Off {
            body["reasoning"] = json!({ "effort": effort_of(l) });
        }
    }
    if let Some(tc) = &req.tool_choice {
        body["tool_choice"] = tool_choice_json(tc);
    }
    let mut tools: Vec<Value> = req.tools.iter().map(function_tool_json).collect();
    tools.extend(req.server_tools.iter().filter_map(server_tool_json));
    if !tools.is_empty() {
        body["tools"] = json!(tools);
    }
    body
}

/// `low` or `high` — xAI's only two reasoning efforts.
///
/// `Medium` rounds **down**, not up: on a cost-weighted metric where output is ~5x input, buying
/// more reasoning than was asked for is the worse error.
fn effort_of(l: ThinkingLevel) -> &'static str {
    match l {
        ThinkingLevel::High => "high",
        _ => "low",
    }
}

/// A client-side tool. Responses declares these **flat** — `{"type":"function","name":…}` — where
/// `chat/completions` nests them under `"function"`.
fn function_tool_json(t: &ToolDef) -> Value {
    json!({
        "type": "function",
        "name": t.name,
        "description": t.description,
        "parameters": t.schema,
    })
}

/// A provider-run tool. `None` for tools this transport does not run; [`negotiate`] refuses those
/// before they reach here, and a test pins the two lists together.
pub(crate) fn server_tool_json(t: &ServerTool) -> Option<Value> {
    Some(match t {
        ServerTool::WebSearch {
            allowed_domains,
            blocked_domains,
            ..
        } => {
            let mut v = json!({ "type": "web_search" });
            // allowed_domains and excluded_domains cannot both be set; allowed wins.
            if !allowed_domains.is_empty() {
                v["filters"] = json!({ "allowed_domains": allowed_domains });
            } else if !blocked_domains.is_empty() {
                v["filters"] = json!({ "excluded_domains": blocked_domains });
            }
            v
        }
        ServerTool::XSearch {
            allowed_handles,
            blocked_handles,
            from_date,
            to_date,
        } => {
            let mut v = json!({ "type": "x_search" });
            if !allowed_handles.is_empty() {
                v["allowed_x_handles"] = json!(allowed_handles);
            } else if !blocked_handles.is_empty() {
                v["excluded_x_handles"] = json!(blocked_handles);
            }
            if let Some(d) = from_date {
                v["from_date"] = json!(d);
            }
            if let Some(d) = to_date {
                v["to_date"] = json!(d);
            }
            v
        }
        ServerTool::CodeExecution => json!({ "type": "code_interpreter" }),
        ServerTool::CollectionsSearch {
            collection_ids,
            limit,
        } => {
            let mut v = json!({ "type": "collections_search", "collection_ids": collection_ids });
            if let Some(n) = limit {
                v["limit"] = json!(n);
            }
            v
        }
        // Not xAI tools; negotiate refuses them before they reach here.
        ServerTool::WebFetch { .. } | ServerTool::UrlContext => return None,
    })
}

fn tool_choice_json(tc: &ToolChoice) -> Value {
    match tc {
        ToolChoice::Auto => json!("auto"),
        ToolChoice::Required => json!("required"),
        ToolChoice::None => json!("none"),
        ToolChoice::Tool(n) => json!({ "type": "function", "name": n }),
    }
}

/// Messages → the `input` array.
///
/// Assistant tool calls and their results round-trip as `function_call` / `function_call_output`
/// items keyed by `call_id` — the id the stream reports as `call_id`, **not** the `fc_…` item id.
fn input_items(msgs: &[Message]) -> Value {
    let mut out: Vec<Value> = Vec::new();
    for m in msgs {
        let role = match m.role {
            Role::User => "user",
            Role::Assistant => "assistant",
        };
        for block in &m.content {
            match block {
                ContentBlock::Text { text, .. } if !text.is_empty() => {
                    out.push(json!({ "role": role, "content": text }));
                }
                ContentBlock::ToolUse { id, name, input } => {
                    out.push(json!({
                        "type": "function_call",
                        "call_id": id,
                        "name": name,
                        "arguments": input.to_string(),
                    }));
                }
                ContentBlock::ToolResult {
                    tool_use_id,
                    content,
                    ..
                } => {
                    out.push(json!({
                        "type": "function_call_output",
                        "call_id": tool_use_id,
                        "output": content,
                    }));
                }
                _ => {}
            }
        }
    }
    json!(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use lvz_protocol::{negotiate, SystemPrompt};

    fn negotiated(req: ChatRequest) -> Negotiated<XaiResponsesCaps> {
        negotiate::<XaiResponsesCaps>(req)
            .1
            .expect("fixture must negotiate")
    }

    fn base() -> ChatRequest {
        let mut r = ChatRequest::new("grok-4.6");
        r.max_tokens = 256;
        r.messages = vec![Message::user("hi")];
        r
    }

    #[test]
    fn the_body_uses_the_responses_field_names_not_chat_completions() {
        let mut r = base();
        r.system = Some(SystemPrompt {
            text: "be terse".into(),
            cache: false,
        });
        let b = build_body(&negotiated(r));
        // `input` not `messages`, `instructions` not a system message, `max_output_tokens` not
        // `max_tokens`. Getting any of these wrong is a 400 from the endpoint.
        assert!(b.get("input").is_some(), "{b}");
        assert!(b.get("messages").is_none(), "{b}");
        assert_eq!(b["instructions"], "be terse");
        assert_eq!(b["max_output_tokens"], 256);
        assert!(b.get("max_tokens").is_none(), "{b}");
        assert_eq!(b["stream"], true);
    }

    #[test]
    fn function_tools_are_declared_flat_not_nested() {
        let mut r = base();
        r.tools = vec![ToolDef {
            name: "read_file".into(),
            description: "read a file".into(),
            schema: json!({"type": "object"}),
            cache: false,
            strict: false,
        }];
        let b = build_body(&negotiated(r));
        let t = &b["tools"][0];
        assert_eq!(t["type"], "function");
        // Flat: the name sits beside `type`, NOT under a "function" object the way
        // chat/completions nests it.
        assert_eq!(t["name"], "read_file");
        assert!(t.get("function").is_none(), "{t}");
    }

    #[test]
    fn tool_calls_round_trip_by_call_id() {
        let mut r = base();
        r.messages = vec![
            Message::user("read it"),
            Message {
                role: Role::Assistant,
                content: vec![ContentBlock::ToolUse {
                    id: "call-1".into(),
                    name: "read_file".into(),
                    input: json!({"path": "a.txt"}),
                }],
            },
            Message {
                role: Role::User,
                content: vec![ContentBlock::ToolResult {
                    tool_use_id: "call-1".into(),
                    content: "contents".into(),
                    is_error: false,
                }],
            },
        ];
        let b = build_body(&negotiated(r));
        let items = b["input"].as_array().expect("input array");
        assert_eq!(items[1]["type"], "function_call");
        assert_eq!(items[1]["call_id"], "call-1");
        // Arguments go over the wire as a JSON *string*, not an object.
        assert!(items[1]["arguments"].is_string(), "{}", items[1]);
        assert_eq!(items[2]["type"], "function_call_output");
        assert_eq!(items[2]["call_id"], "call-1");
        assert_eq!(items[2]["output"], "contents");
    }

    #[test]
    fn medium_thinking_rounds_down_to_low() {
        // xAI has only `low` and `high`. On a cost-weighted metric where output is ~5x input,
        // buying more reasoning than was asked for is the worse error.
        for (level, want) in [
            (ThinkingLevel::Low, "low"),
            (ThinkingLevel::Medium, "low"),
            (ThinkingLevel::High, "high"),
        ] {
            let mut r = base();
            r.thinking = Some(level);
            let b = build_body(&negotiated(r));
            assert_eq!(b["reasoning"]["effort"], want, "{level:?}");
        }
        // Off emits no reasoning block at all.
        let mut r = base();
        r.thinking = Some(ThinkingLevel::Off);
        assert!(build_body(&negotiated(r)).get("reasoning").is_none());
    }

    #[test]
    fn web_search_domain_filters_are_mutually_exclusive() {
        // allowed_domains and excluded_domains cannot both be set; allowed wins.
        let v = server_tool_json(&ServerTool::WebSearch {
            max_uses: None,
            allowed_domains: vec!["docs.rs".into()],
            blocked_domains: vec!["spam.example".into()],
        })
        .expect("web_search maps");
        assert_eq!(v["filters"]["allowed_domains"][0], "docs.rs");
        assert!(v["filters"].get("excluded_domains").is_none(), "{v}");

        let v = server_tool_json(&ServerTool::WebSearch {
            max_uses: None,
            allowed_domains: vec![],
            blocked_domains: vec!["spam.example".into()],
        })
        .unwrap();
        assert_eq!(v["filters"]["excluded_domains"][0], "spam.example");
    }

    /// The declared capability list and what `server_tool_json` actually maps must agree.
    ///
    /// Widening the mapper without widening the declaration would let a tool through that the
    /// provider never receives; widening the declaration without the mapper would accept a tool
    /// and then drop it. Either way the model is offered something that does not work, silently.
    #[test]
    fn the_declaration_and_the_mapper_agree() {
        let every = [
            (
                ServerTool::WebSearch {
                    max_uses: None,
                    allowed_domains: vec![],
                    blocked_domains: vec![],
                },
                Capability::WebSearch,
            ),
            (
                ServerTool::WebFetch { max_uses: None },
                Capability::WebFetch,
            ),
            (ServerTool::CodeExecution, Capability::CodeExecution),
            (
                ServerTool::XSearch {
                    allowed_handles: vec![],
                    blocked_handles: vec![],
                    from_date: None,
                    to_date: None,
                },
                Capability::XSearch,
            ),
            (
                ServerTool::CollectionsSearch {
                    collection_ids: vec!["c".into()],
                    limit: None,
                },
                Capability::CollectionsSearch,
            ),
            (ServerTool::UrlContext, Capability::UrlContext),
        ];
        let declared = XaiResponsesCaps::declare();
        for (tool, cap) in every {
            let maps = server_tool_json(&tool).is_some();
            assert_eq!(
                maps,
                declared.supports(cap),
                "{cap:?}: mapper says {maps}, declaration says {}",
                declared.supports(cap)
            );
        }
    }
}
